packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts

```diff
@@ -444,20 +315,41 @@ export class ModuleUrlRebasingImporter extends RelativeUrlRebasingImporter {
   constructor(
     entryDirectory: string,
     directoryCache: Map<string, DirectoryEntry>,
+    private moduleCache: Map<string, ModuleEntry>,
     rebaseSourceMaps: Map<string, RawSourceMap> | undefined,
-    private finder: FileImporter<'sync'>['findFileUrl'],
+    private finder: (
+      specifier: string,
+      options: { fromImport: boolean; resolveDir?: string },
+    ) => URL | null,
   ) {
     super(entryDirectory, directoryCache, rebaseSourceMaps);
   }
 
   override canonicalize(url: string, options: { fromImport: boolean }): URL | null {
-    if (url.startsWith('file://')) {
+    if (url[0] === 'f' && url.startsWith('file://')) {
       return super.canonicalize(url, options);
     }
 
-    const result = this.finder(url, options);
+    let resolveDir;
+    if (url[0] === 'n' && url.startsWith('ng-package;')) {
+      const values = url.split(';', 3);
+      resolveDir = values[1];
+      url = values[2];
+    }
+
+    let result: URL | undefined | null = this.moduleCache.get(url)?.resolved;
```

Comment:
The module cache seems kind of unrelated to this rebaser/canonicalizer. I wonder if we could pull that out to a slightly higher level of abstraction and maybe cache before/after calling `canonicalize()` to separate concerns a little cleaner?

That would probably apply the cache to more than just `ng-package;` specifiers, but maybe that's a good thing? Not sure how viable this is, just a thought which stood out to me.

Reply:
Currently it would apply to all non-relative resolution attempts.  Since we don't control the execution of `canonicalize` it needs to be done inside. `canonicalize` is called by the Sass compiler itself and is akin to a onResolve hook.

I dropped the cache from this PR.  It does result in less of a performance improvement overall here but it can be introduced in a separate change. I wanted to eventually make the cache work across all active Sass workers anyway.